### PR TITLE
Compute loss independent from decoder for TF EncDec models (as #14139)

### PR DESCRIPTION
**Pull request by ydshieh**
# What does this PR do?

Apply the same change in #14139 to TF Encoder-Decoder like models.

(As discussed in [this comment in 14148](https://github.com/huggingface/transformers/pull/14148#discussion_r780361117))

@NielsRogge & @patrickvonplaten  knows better the context
@gante @Rocketknight1 for TF

### CODE REVIEW

File: tests/test_modeling_tf_encoder_decoder.py [hunk 1]
```diff
@@ -240,7 +240,7 @@ def check_encoder_decoder_model_labels(
         assert "loss" in outputs_encoder_decoder
 
         batch_size, seq_len = decoder_input_ids.shape
-        expected_shape = (batch_size, seq_len - 1, decoder_config.vocab_size)
+        expected_shape = (batch_size, seq_len, decoder_config.vocab_size)
```
Comment (ydshieh) [inline tests/test_modeling_tf_encoder_decoder.py:243]: This is changed, because in TF Causal LM models, the returned logits is the truncated one `logits = logits[:, :-1]`.
However, in PyTorch Causal LM models, the returned logits is 

```
        prediction_scores = self.cls(sequence_output)

        if labels is not None:
            # we are doing next-token prediction; shift prediction scores and input ids by one
            shifted_prediction_scores = prediction_scores[:, :-1, :].contiguous()
            ...
            lm_loss = loss_fct(shifted_prediction_scores.view(-1, self.config.vocab_size), labels.view(-1))

        return CausalLMOutputWithCrossAttentions(
            loss=lm_loss,
            logits=prediction_scores,
            ...
```
When using the same logic as in #14139 to compute loss independent from decoder, there is no truncation needed to be done.

File: src/transformers/models/encoder_decoder/modeling_tf_encoder_decoder.py [hunk 2]
```diff
@@ -40,6 +47,13 @@
 
 _CONFIG_FOR_DOC = "EncoderDecoderConfig"
 
+DEPRECATION_WARNING = (
+    "Version v4.12.0 introduces a better way to train encoder-decoder models by computing the loss inside the "
```
Comment (NielsRogge) [inline src/transformers/models/encoder_decoder/modeling_tf_encoder_decoder.py:51]:
```suggestion
    "Version v4.17.0 introduces a better way to train encoder-decoder models by computing the loss inside the "
```

File: src/transformers/models/encoder_decoder/modeling_tf_encoder_decoder.py [hunk 3]
```diff
@@ -40,6 +47,13 @@
...
+    "encoder-decoder framework rather than in the decoder itself. You may observe training discrepancies if fine-tuning "
+    "a model trained with versions anterior to 4.12.0. The decoder_input_ids are now created based on the labels, no "
```
Comment (NielsRogge) [inline src/transformers/models/encoder_decoder/modeling_tf_encoder_decoder.py:53]:
```suggestion
    "a model trained with versions anterior to 4.17.0. The decoder_input_ids are now created based on the labels, no "
```

File: src/transformers/models/vision_encoder_decoder/modeling_tf_vision_encoder_decoder.py [hunk 4]
```diff
@@ -40,6 +47,13 @@
 
 _CONFIG_FOR_DOC = "VisionEncoderDecoderConfig"
 
+DEPRECATION_WARNING = (
+    "Version v4.12.0 introduces a better way to train encoder-decoder models by computing the loss inside the "
```
Comment (NielsRogge) [inline src/transformers/models/vision_encoder_decoder/modeling_tf_vision_encoder_decoder.py:51]:
```suggestion
    "Version v4.17.0 introduces a better way to train encoder-decoder models by computing the loss inside the "
```

File: src/transformers/models/vision_encoder_decoder/modeling_tf_vision_encoder_decoder.py [hunk 5]
```diff
@@ -40,6 +47,13 @@
...
+    "encoder-decoder framework rather than in the decoder itself. You may observe training discrepancies if fine-tuning "
+    "a model trained with versions anterior to 4.12.0. The decoder_input_ids are now created based on the labels, no "
```
Comment (NielsRogge) [inline src/transformers/models/vision_encoder_decoder/modeling_tf_vision_encoder_decoder.py:53]:
```suggestion
    "a model trained with versions anterior to 4.17.0. The decoder_input_ids are now created based on the labels, no "
```

**Review by NielsRogge — APPROVED** [review]
Thanks for working on this!

**Review by gante — APPROVED** [review]
LGTM 👍 Thanks for fixing, @ydshieh

File: src/transformers/models/vision_encoder_decoder/modeling_tf_vision_encoder_decoder.py [hunk 6]
```diff
@@ -134,8 +148,36 @@
 """
 
 
+def shift_tokens_right(input_ids: tf.Tensor, pad_token_id: int, decoder_start_token_id: int):
```
[the remark this answers is not in the document]
Comment (ydshieh) [reply src/transformers/models/vision_encoder_decoder/modeling_tf_vision_encoder_decoder.py:152]: Very good point @gante, thanks!
